src/Type/Php/HashHmacFunctionDynamicReturnTypeExtension.php

```diff
@@ -0,0 +1,64 @@
+<?php declare(strict_types = 1);
+
+namespace PHPStan\Type\Php;
+
+use PhpParser\Node\Expr\FuncCall;
+use PHPStan\Analyser\Scope;
+use PHPStan\Php\PhpVersion;
+use PHPStan\Reflection\FunctionReflection;
+use PHPStan\Type\Accessory\AccessoryNonEmptyStringType;
+use PHPStan\Type\Constant\ConstantBooleanType;
+use PHPStan\Type\Constant\ConstantStringType;
+use PHPStan\Type\DynamicFunctionReturnTypeExtension;
+use PHPStan\Type\NeverType;
+use PHPStan\Type\Type;
+use PHPStan\Type\TypeCombinator;
+use function count;
+use function hash_hmac_algos;
+use function in_array;
+
+class HashHmacFunctionDynamicReturnTypeExtension implements DynamicFunctionReturnTypeExtension
+{
+
+	public function __construct(private PhpVersion $phpVersion)
+	{
+	}
+
+	public function isFunctionSupported(FunctionReflection $functionReflection): bool
+	{
+		return $functionReflection->getName() === 'hash_hmac';
+	}
+
+	public function getTypeFromFunctionCall(FunctionReflection $functionReflection, FuncCall $functionCall, Scope $scope): ?Type
+	{
+		$args = $functionCall->getArgs();
+
+		if (count($args) < 3) {
+			return null;
+		}
+
+		$algo = $scope->getType($args[0]->value);
+
+		if (!$algo instanceof ConstantStringType) {
```

Comment:
you shouldn't use `instanceof ConstantStringType` as it is [deprecated](https://phpstan.org/blog/why-is-instanceof-type-wrong-and-getting-deprecated). The [article](https://phpstan.org/blog/why-is-instanceof-type-wrong-and-getting-deprecated) goes into details about what you should do instead.

This also forces handling `$algos` being a union which improves this extension overall